File: src/Applications/Orbital.Web/Orbital.Forecast.BlazorApplication/BlazorApp.Client/Library/DynamicLinkLibraries/Motion6D/Motion6D.Interfaces/Comparers/PositionComparer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Motion6D.Interfaces.Comparers
{
    class PositionComparer : IComparer<IPosition>
    {
        #region Fields

        static internal readonly PositionComparer Singleton = new PositionComparer();

        #endregion

        #region IComparer<IPosition> Members

        int IComparer<IPosition>.Compare(IPosition x, IPosition y)
        {
            if (IsSource(x, y))
            {
                return -1;
            }
            if (IsSource(y, x))
            {
                return 1;
            }
            return 0;
        }

        #endregion

        #region Private Members

        bool IsSource(IPosition source, IPosition target)
        {
            if (target.Parent == null)
            {
                return false;
            }
            if (target.Parent == source)
            {
                return true;
            }
            if (IsSource(source, target.Parent))
            {
                return true;
            }
            return false;
        }

        #endregion
    }
}

File: src/TypeScript/TypeScriptBackup/src/Library/Motion6D/Comparators/PositionComparer.ts
import type { IComparator } from "../../Utilities/Sort/Interfaces/IComparator";
import type { IPosition } from "../Interfaces/IPosition";

export class PositionComparer implements IComparator<IPosition> {
    compare(x: IPosition, y: IPosition): number {
        if (this.isSource(x, y)) return -1;
        if (this.isSource(y, x)) return 1;
        return 0;

    }
    isSource(source: IPosition, target: IPosition) {
        var tp = target.getParentFrame()
        if (tp === undefined) {
            return false;
        }
        if (tp == source) {
            return true;
        }
        if (this.isSource(source, tp)) {
            return true;
        }
        return false;
    }
}